### ISSUE TITLE
Error span for missing main function is too large

### ISSUE THREAD

**Issue by lnicola** [rust-lang/rust#77968]
I tried this code:

```rust
fn foo() {}
fn bar() {}
fn baz() {}
```

I expected to see this happen:

```
error[E0601]: `main` function not found in crate `hello`
 --> src/main.rs:4:1
3 | | fn baz() {}
4 | ^ consider adding a `main` function to `src/main.rs`
```

(or something similar -- an empty last line might be missing).

Instead, this happened:

```
error[E0601]: `main` function not found in crate `hello`
 --> src/main.rs:1:1
  |
1 | / fn foo() {}
2 | | fn bar() {}
3 | | fn baz() {}
  | |___________^ consider adding a `main` function to `src/main.rs`
```

This matters for IDE support, because using the whole file as an error span is a bit messy, marking potentially hundreds of lines as an error.

### Meta

`rustc --version`:
```
rustc 1.49.0-nightly (91a79fb29 2020-10-07)
```

**Comment by GopherJ** [comment]
As you can see from the below screenshot, `}` is missing in line 4, but it mess up all:)

![image](https://user-images.githubusercontent.com/33961674/98039568-2ea02a00-1e1f-11eb-8e75-560b6c063db1.png)

**Comment by lnicola** [comment]
@GopherJ that's a `rust-analyzer` parser recovery issue, completely unrelated to this one.

**Comment by dtolnay** [comment]
> I don't know when I'll be able to return to this. Closing for now. https://github.com/rust-lang/rust/pull/79086#issuecomment-773020629

@rustbot release-assignment

**Comment by euclio** [comment]
Looks like this was fixed by #93142.

**Comment by dtolnay** [comment]
Indeed! Thanks for finding that.
